api/api_ethereum.go

```diff
@@ -1429,13 +1445,62 @@ func EthDoEstimateGas(ctx context.Context, b Backend, args EthTransactionArgs, b
 	}
 	// Reject the transaction as invalid if it still fails at the highest allowance
 	if hi == cap {
-		isExecutable, err := executable(hi)
+		isExecutable, ret, vmErr, err := executable(hi)
 		if err != nil {
 			return 0, err
 		}
 		if !isExecutable {
-			return 0, fmt.Errorf("gas required exceeds allowance or always failing transaction")
+			if ret != nil {
+				if vmErr == vm.ErrExecutionReverted || strings.Contains(vmErr.Error(), vm.ErrExecutionReverted.Error()) {
+					return 0, newRevertError(ret)
+				}
+				return 0, vmErr
+			}
+			// Otherwise, the specified gas cap is too low
+			return 0, fmt.Errorf("gas required exceeds allowance (%d)", cap)
 		}
 	}
 	return hexutil.Uint64(hi), nil
 }
+
+// isVMError returns true if given error is occurred during EVM execution.
+func isVMError(err error) bool {
+	for _, vmError := range vmErrors {
+		if err == vmError || strings.Contains(err.Error(), vmError.Error()) {
+			return true
+		}
+	}
+	return false
+}
+
+// newRevertError wraps data returned when EVM execution was reverted.
+// Make sure that data is returned when execution reverted situation.
+func newRevertError(data []byte) *revertError {
+	reason, errUnpack := abi.UnpackRevert(data)
+	err := errors.New("execution reverted")
+	if errUnpack == nil {
+		err = fmt.Errorf("execution reverted: %v", reason)
+	}
+	return &revertError{
+		error:  err,
+		reason: hexutil.Encode(data),
+	}
+}
+
+// revertError is an API error that encompassas an EVM revertal with JSON error
+// code and a binary data blob.
+type revertError struct {
```

Comment:
Don't we need to implement `Error() string` for this?

Reply:
@jimni1222 
I'll update this PR soon. I should have checked via http call, not only via console. 
Thanks a lot.

Reply:
I applied that in this [commit](https://github.com/klaytn/klaytn/pull/1134/commits/2769453baa2c0b2d2d73e07602004bc5efdfdd63).

After this commit, we can get exactly same return format with Geth.
**KEN:**
```shell
$ curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "method": "eth_estimateGas", "params": [{"from": "0x3753f5bf8cCA929EeeB4B74F200dfc6375ad5444", "to": "0xbe3892d33620be5aca8c75d39e7401871194d290", "data": "0xa6f9dae1000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d"}], "id": 1}' http://localhost:8551
{"jsonrpc":"2.0","id":1,"error":{"code":3,"message":"execution reverted: Caller is not owner","data":"0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001343616c6c6572206973206e6f74206f776e657200000000000000000000000000"}}
```

**Geth:**
```shell
$ curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "method": "eth_estimateGas", "params": [{"from": "0xca7a99380131e6c76cfa622396347107aeedca2d", "to": "0x27a110C35BA478034d7F874C89787E0FE587f65F", "data": "0xa6f9dae1000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d"}], "id": 1}' http://localhost:8545
{"jsonrpc":"2.0","id":1,"error":{"code":3,"message":"execution reverted: Caller is not owner","data":"0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001343616c6c6572206973206e6f74206f776e657200000000000000000000000000
```

Related Ethereum's PR is https://github.com/ethereum/go-ethereum/pull/22677

Reply:
@aeharvlee I think we need to implement in klay namspace later. Could you make a task issue for v1.9.0?